src/libs/logging/src/file_rotate.rs

```diff
@@ -0,0 +1,266 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+use std::fs::{self, File, OpenOptions};
+use std::io::{self, LineWriter, Result, Write};
+use std::path::{Path, PathBuf};
+use std::time::{Duration, Instant};
+
+// Default rotate size for logger files.
+fn default_rotate_size() -> u64 {
+    10485760
+}
+
+// Default number of log files to keep.
+fn default_rotate_keep() -> usize {
+    3
+}
+
+/// Writer with file rotation for log files.
+///
+/// This is a modified version of `FileAppender` from
+/// https://github.com/sile/sloggers/blob/153c00a59f7218c1d96f522fb7a95c80bb0d530c/src/file.rs#L190
+#[derive(Debug)]
+pub struct FileRotator {
+    path: PathBuf,
+    file: Option<LineWriter<File>>,
+    truncate: bool,
+    written_size: u64,
+    rotate_size: u64,
+    rotate_keep: usize,
+    reopen_check_interval: Duration,
+    next_reopen_check: Instant,
+}
+
+impl FileRotator {
+    /// Create a new instance of [`FileRotator`] to write log file at `path.
+    ///
+    /// Caller needs to ensure the parent directory already exists.
+    pub fn new<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let p = Path::new(path.as_ref());
+        if let Ok(md) = p.metadata() {
+            if !md.is_file() {
+                return Err(io::Error::from(io::ErrorKind::InvalidInput));
+            }
+        }
+        if let Some(parent) = p.parent() {
+            if p.has_root() || !parent.as_os_str().is_empty() {
+                let md = parent.metadata()?;
+                if !md.is_dir() {
+                    return Err(io::Error::from(io::ErrorKind::InvalidInput));
+                }
+            }
+        }
+
+        Ok(FileRotator {
+            path: path.as_ref().to_path_buf(),
+            file: None,
+            truncate: false,
+            written_size: 0,
+            rotate_size: default_rotate_size(),
+            rotate_keep: default_rotate_keep(),
+            reopen_check_interval: Duration::from_millis(1000),
+            next_reopen_check: Instant::now(),
+        })
+    }
+
+    /// Enable/disable log file truncation.
+    pub fn truncate(&mut self, truncate: bool) -> &mut Self {
+        self.truncate = truncate;
+        self
+    }
+
+    /// Set the threshold size to rotate log files.
+    pub fn rotate_threshold(&mut self, size: u64) -> &mut Self {
+        self.rotate_size = size;
+        self
+    }
+
+    /// Set number of rotated log files to keep.
+    pub fn rotate_count(&mut self, count: usize) -> &mut Self {
+        self.rotate_keep = count;
+        self
+    }
+
+    /// Set the interval to check for rotation.
+    pub fn check_interval(&mut self, interval: Duration) -> &mut Self {
+        self.reopen_check_interval = interval;
+        self
+    }
+
+    fn reopen_if_needed(&mut self) -> Result<()> {
```

Comment:
I think the more complex functions would benefit from doc comments.

Reply:
How about this:
```
    // Open the log file if
    // - it hasn't been open yet.
    // - the log previous log file has been rotated as history log file and a new one is needed.
```